obs-studio-server/source/memory-manager.cpp

```diff
@@ -33,106 +32,248 @@ MemoryManager::MemoryManager()
+void MemoryManager::calculateRawSize(source_info* si)
 {
-	mtx.lock();
-	if (strcmp(obs_source_get_id(source), "ffmpeg_source") != 0) {
-		mtx.unlock();
-		return;
+	calldata_t      cd = {0};
+	proc_handler_t* ph = obs_source_get_proc_handler(si->source);
+	proc_handler_call(ph, "get_nb_frames", &cd);
+
+	uint64_t nb_frames = calldata_int(&cd, "num_frames");
+	uint64_t width     = calldata_int(&cd, "width");
+	uint64_t height    = calldata_int(&cd, "height");
+	uint32_t pix_fmt   = calldata_int(&cd, "pix_format");
+	float    bpp       = 0;
+
+	switch (pix_fmt) {
+		case VIDEO_FORMAT_I420:
+		case VIDEO_FORMAT_NV12:
+			bpp = 1.5;
+			break;
+		case VIDEO_FORMAT_YVYU:
+		case VIDEO_FORMAT_YUY2:
+		case VIDEO_FORMAT_UYVY:
+			bpp = 2;
+			break;
+		case VIDEO_FORMAT_RGBA:
+		case VIDEO_FORMAT_BGRA:
+		case VIDEO_FORMAT_BGRX:
+		case VIDEO_FORMAT_Y800:
+			bpp = 4;
+			break;
+		case VIDEO_FORMAT_I444:
+			bpp = 4;
+			break;
 	}
 
-	source_info* info = new source_info;
-	info->cached      = false;
-	info->size        = 0;
-	info->source      = source;
-	sources.emplace(obs_source_get_name(source), info);
+	uint64_t size = width * height * bpp * nb_frames;
+	si->size = size < 0 ? 0 : size;
+}
+
+bool MemoryManager::shouldCacheSource(source_info* si)
+{
+	bool should_cache = false;
+
+	obs_data_t* settings = obs_source_get_settings(si->source);
+
+	bool looping        = obs_data_get_bool(settings, "looping");
+	bool local_file     = obs_data_get_bool(settings, "is_local_file");
+	bool enable_caching = config_get_bool(ConfigManager::getInstance().getGlobal(), "General", "fileCaching");
+	bool is_small       = current_cached_size + si->size < allowed_cached_size;
+	bool showing        = obs_source_showing(si->source);
+
+	if (!showing && !obs_data_get_bool(settings, "close_when_inactive"))
+		showing = true;
 
-	mtx.unlock();
-	updateCacheSettings(source, false);
+	obs_data_release(settings);
+
+	return looping &&
+           local_file &&
+           enable_caching &&
+           is_small && showing;
 }
 
-void MemoryManager::updateCacheSettings(obs_source_t* source, bool updateSize)
+void updateSource(obs_source_t* source, bool caching)
 {
-	std::unique_lock<std::mutex> ulock(mtx);
+	obs_data_t* settings = obs_source_get_settings(source);
+	if (obs_data_get_bool(settings, "caching") != caching) {
+		obs_data_set_bool(settings, "caching", caching);
+		obs_source_update(source, settings);
+	}
+	obs_data_release(settings);
+}
 
-	auto it = sources.find(obs_source_get_name(source));
-	if (it == sources.end())
+void MemoryManager::addCachedMemory(source_info* si)
+{
+	std::unique_lock<std::mutex> ulock(si->mtx);
+
+	if (!si->size || si->cached || current_cached_size + si->size > allowed_cached_size)
 		return;
 
-	obs_data_t* settings = obs_source_get_settings(source);
+	blog(LOG_INFO, "adding %dMB, source: %s", si->size / 1000000, obs_source_get_name(si->source));
+	current_cached_size += si->size;
```

Comment:
If this can be called by multiple threads, shouldn't it be protected by a global mutex since it belongs to the memory manager singleton?

Reply:
Seems like there is one missing in the sourceManager thread, but it is currently protected inside the monitorMemory thread, unregister function. Will add it there too 👍